tools/generators/lib/PBXProj/src/Identifiers.swift

```diff
@@ -81,6 +83,161 @@ FF01000000000000000001\#(String(format: "%02X", index)) \#
+        ///   - path: The file path the build file is referencing.
+        ///   - type: The type of build file being identified.
+        ///   - shard: The generator shard the target belongs to.
+        ///   - hashCache: A cache that will be used to guarantee that the
+        ///     sub-identifier returned is unique.
+        public static func subIdentifier(
+            _ path: BazelPath,
+            type: FileType,
+            shard: UInt8,
+            hashCache: inout [UInt8: Set<String>]
+        ) -> SubIdentifier {
+            precondition(shard < 0xFE)
+            return SubIdentifier(
+                shard: String(format: "%02X", shard),
+                type: type,
+                path: path,
+                hash: shardSubIdentifier(
+                    path.path + (path.isFolder ? "0" : "1"),
+                    hashCache: &hashCache[shard, default: []]
+                )
+            )
+        }
+
+        public static func compileStubSubIdentifier(
+            targetSubIdentifier: Targets.SubIdentifier
+        ) -> SubIdentifier {
+            // We purposely store less information here than in
+            // `subIdentifier()`, to reduce the amount of data written into
+            // BuildFileMaps
+            return SubIdentifier(
+                shard: targetSubIdentifier.shard,
+                type: .compileStub,
+                path: BazelPath(""),
+                hash: targetSubIdentifier.hash
+            )
+        }
+
+        public static func id(
+            subIdentifier: SubIdentifier
+        ) -> String {
+            let type = subIdentifier.type
+
+            switch type {
+            case .compileStub:
+                return #"""
+\#(subIdentifier.shard)00\#(subIdentifier.hash)0000000000FF \#
+/* _CompileStub_.m in Sources */
+"""#
+
+            default:
+                let basename = subIdentifier.path.path
+                    .split(separator: "/").last!
+                return #"""
+\#(subIdentifier.shard)FF\#(subIdentifier.hash) \#
+/* \#(basename) in \#(type.buildPhase.name) */
+"""#
+            }
+        }
+
+        private static func shardSubIdentifier(
+            _ hashable: String,
+            hashCache: inout Set<String>
+        ) -> String {
+            var hash: String
+            var retryCount = 0
+            repeat {
+                hash = buildFileHash(hashable, retryCount: retryCount)
+                retryCount += 1
+            } while hashCache.contains(hash)
+
+            hashCache.insert(hash)
+
+            return hash
+        }
+
+        private static func buildFileHash(
+            _ hashable: String,
+            retryCount: Int
+        ) -> String {
+            let content: String
+            if retryCount == 0 {
+                content = hashable
+            } else {
+                content = "\(hashable)\0\(retryCount)"
+            }
+
+            let digest = Insecure.MD5.hash(data: content.data(using: .utf8)!)
```

Comment:
This should let you avoid the force unwrap

```suggestion
            let digest = Insecure.MD5.hash(data: Data(content.utf8))
```